File: cpp/0131分割回文串.cpp
#include<vector>
#include<iostream>

using namespace std;

vector<vector<string>> res;
vector<string> temp;


bool IsPalindrome(string &s,int left,int right)
{
    for (int i = 0; i <= ((right-left)>>1);++i)
        if(s[i+left]!=s[right-i])
            return false;
    return true;
}

void BackTracking(int start,string &s)
{
    if (start >= s.size())
    {
        res.push_back(temp);
        return;
    }

    for (int i = start;i<s.size();++i)
    {
        
        if (IsPalindrome(s, start, i))
        {
            
            temp.push_back(s.substr(start, i - start + 1));
            BackTracking(i+1, s);
            temp.pop_back();
        }

        else
            continue;
    }

}

vector<vector<string>> partition(string s)
{
    BackTracking(0, s);
    return res;
}


int main()
{
    partition("aab");
}

File: python/main/0131分割回文串.py


def Partition(s: str) -> list:
    res=[]
    temp=[]

    def _isPalindrome(left:int,right:int):
        nonlocal s
        
        while(left<right):
            if(s[left]!=s[right]):
                return False
            left+=1
            right-=1
        
        return True

    def _backTracking(start:int):
        nonlocal s
        
        if(start>=len(s)):
            res.append(temp.copy())
            return
        
        for i in range(start,len(s)):
            if(_isPalindrome(start,i)):
                temp.append(s[start:i+1])
                _backTracking(i+1)
                temp.pop()
            
    _backTracking(0)
    return res


print(Partition("aab"))
# print("aab"[0:0])